src/dune_pkg/rev_store.ml

```diff
@@ -301,55 +339,188 @@ module File = struct
         match Re.Group.get m 1 with
         | "blob" ->
           Some
-            { hash = Re.Group.get m 2
-            ; size = Int.of_string_exn @@ Re.Group.get m 3
-            ; path = Path.Local.of_string @@ Re.Group.get m 4
-            }
+            (Blob
+               { hash = Re.Group.get m 2
+               ; size = Int.of_string_exn @@ Re.Group.get m 3
+               ; path = Path.Local.of_string @@ Re.Group.get m 4
+               })
+        | "commit" ->
+          Some
+            (Commit
+               { rev = Rev (Re.Group.get m 2)
+               ; path = Path.Local.of_string @@ Re.Group.get m 4
+               })
         | _ -> None)
   ;;
 
-  let path t = t.path
+  let path = function
+    | Blob { path; _ } | Commit { path; _ } -> path
+  ;;
 end
 
+(* a submodule in [.gitmodules] can also have a [branch] but given we only need
+   to resolve the commit
+   object, we don't have to care about the tracking branch *)
+type submodule =
+  { path : Path.Local.t
+  ; source : string
+  }
+
+let parse_submodules =
+  let path_re = Re.(compile @@ seq [ str "path = "; group (rep1 any) ]) in
+  let url_re = Re.(compile @@ seq [ str "url = "; group (rep1 any) ]) in
+  fun gitmodules ->
+    let lines = String.split ~on:'\n' gitmodules in
+    let parse_section lines =
+      let section, rest =
+        List.split_while lines ~f:(fun line -> not @@ String.is_prefix ~prefix:"[" line)
+      in
+      let locate_by_re re line =
+        Re.exec_opt re line |> Option.map ~f:(fun m -> Re.Group.get m 1)
+      in
+      let path = List.find_map section ~f:(locate_by_re path_re) in
+      let url = List.find_map section ~f:(locate_by_re url_re) in
+      match path, url with
+      | Some path, Some source ->
+        let path = Path.Local.of_string path in
+        { path; source }, rest
+      | _, _ -> failwith "Submodule definition missing path & url"
+    in
+    let rec parse submodules = function
+      | [] -> submodules
+      | line :: lines ->
+        (match String.is_prefix ~prefix:"[submodule" line with
```

Comment:
The config parser had a few bugs with respect to parsing, can we switch to something a bit more principled: E.g.

```
type section =
  { name : string
  ; arg : string option
  ; bindings : bindings
  }
val parse : string -> section list
```

Btw, I find it a bit strange that for remotes we chose to use `git remote` to get the information, but for submodules we're parsing `.gitmodules`. Might as well stick do the same thing for both (I'd recommend just parsing the config format, which is more probably more stable than the output of git commands)

Reply:
I believe I had a branch where I replaced a number of git calls with parsing the config files, but if I remember it correctly there was some concern about replacing git commands with parsing the files. Unfortunately I can't seem to find the discussion, @emillon do you remember this? However, which commands are plumbing is not well defined.

Anyway, yes to a simple parser, we seem to need this every now and then. Replacing `git remote` would however be a different PR.

Reply:
Yes, there's no need to touch `git remote` here. I'm just questioning why we're not handling all the metadata consistently.

I don't know why Etienne prefers to use `git remote`, but it seems worse than reading the config file. Git's config files maintain at least a certain level of backwards compatibility b/c newer versions of git need to be able to read them. While the output of individual git commands is not guaranteed to be stable anywhere (although in practice it's quite stable)

Reply:
my opinion was that it's not dune's job to read other programs' config files (at one point, we said we wouldn't parse `.ocamlformat` files for example). but I'm pretty far from this area of the code, and I don't know what guarantees `git` makes for its output and for its config file, so I don't have a strong opinion on this.

Reply:
Avoiding depending on the implementations of external projects is certainly desirable, but there's no advantage on depending on the output of an external command vs the format of a file. There are no stability guarantees on the output of the command as far as I know.

Reply:
what about accessing this data through `git config`? you can list the bindings using `git config --list`, and get one using for example `git config --get remote.origin.url`? I'm pretty sure that this is easier to parse than git config files (and you don't have to know the logic of where to look for files, etc)

Reply:
Wasn't aware of it. Seems like a good option as well. Can it be used to edit the config as well? Can we use it to access submodules?

If yes, seems like this could have saved us some parsing work. The parsing is already done, so it probably wouldn't be our first priority to rewrite it though.

Reply:
Yes, this can be used to edit the config files as well. It returns precise exit codes too (see `git-config(1)`), so I think it would be a useful tool to use.

Reply:
and submodules?

Reply:
do you mean, how to parse `.gitmodules`? yes, you can do that using `git config --file .gitmodules --list` for example. you can also refer to a blob like `--blob main:.gitmodules` if you don't have a working tree.

Reply:
Very cool. Seems like that's all we need then.